samples/vertex-ffp/main.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * This sample provides a very basic demonstration of 3D rendering on the Xbox,
+ * using pbkit. Based on the pbkit demo sources.
+ */
+#include <hal/video.h>
+#include <hal/xbox.h>
+#include <math.h>
+#include <pbkit/pbkit.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <hal/debug.h>
+
+#include <xgu/xgu.h>
+#include <xgu/xgux.h>
+
+#include "xgu_extra.h"
+#include "xgux_extra.h"
+
+//FIXME: Replace by glm or something?
+#include "../common/matrix.h"
+
+#define MAX_RAM 0x3ffb000
+#define ARRAY_SIZE(x) (sizeof(x)/sizeof(x[0]))
+
+static float     m_viewport[4][4];
+
+typedef struct {
+    float pos[3];
+    float color[3];
+} __attribute__((packed)) ColoredVertex;
+
+static const ColoredVertex verts[] = {
+    //  X     Y     Z       R     G     B
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}}, /* Background triangle 1 */
+    {{-1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}}, /* Background triangle 2 */
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0, -1.0,  1.0}, { 1.0,  0.0,  0.0}}, /* Foreground triangle */
+    {{ 0.0,  1.0,  1.0}, { 0.0,  1.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.0,  0.0,  1.0}},
+};
+static ColoredVertex *alloc_vertices;
+
+/* Main program function */
+int main(void)
+{
+    uint32_t *p;
+    int       start, last, now;
+    int       fps, frames, frames_total;
+
+    /* Set display mode */
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    /* Start GPU */
+    int status = pb_init();
+    if (status) {
+        debugPrint("pb_init Error %d\n", status);
+        Sleep(2000);
+        return 1;
+    }
+
+    //FIXME: Remove.
+    //       This works around a race-condition in XQEMU.
+    //       It looks like pb_init uses `debugPrint()` so the GPU and CPU draw
+    //       at the same time. Depending on the order of operations, this
+    //       breaks the XQEMU surface cache.
+    Sleep(100);
+
+    /* Display GPU result on display */
+    pb_show_front_screen();
+
+    /* Basic setup */
+    int width = pb_back_buffer_width();
+    int height = pb_back_buffer_height();
+
+    /* Setup fragment shader */
+    p = pb_begin();
+    #include "ps.inl"
+    pb_end(p);
+
+    /* Load constant geometry */
+    alloc_vertices = MmAllocateContiguousMemoryEx(sizeof(verts), 0, MAX_RAM, 0, PAGE_READWRITE | PAGE_WRITECOMBINE);
+    memcpy(alloc_vertices, verts, sizeof(verts));
+
+    /* Setup viewport matrix */
+    //FIXME: This matrix should be transposed
+    matrix_viewport(m_viewport, 0, 0, width, height, 0, (float)0xFFFF);
+
+    //FIXME: This is already set in pbkit probably?
+    //p = xgu_set_viewport_offset(p, 0.0f, 0.0f, 0.0f, 0.0f);
+    //p = xgu_set_viewport_scale(p, 1.0f / width, 1.0f / height, 1.0f / (float)0xFFFF, 1.0f);
```

Comment:
I don't think this is done in pbkit.
By the looks of it, I'd have to use a scale of `1.0f`.
The offset would be `-width/2.0f, -height/2.0f, 0.0f, 0.0f` or something.

Reply:
**Note: This old code was also missing `pb_begin` / `pb_end`. So it was probably never effective.**

However, this is exactly where the difference lies. Setting offset to all 0.0 makes this look good on hardware (as offset is already part of viewport matrix). Scale seems to be ignored by hardware. I'm not sure when it would be used.

I have not experimented with Z and W, so I was unable to make this work properly in XQEMU (no issue created yet; although one might exist already?). I tried to modify `inv_viewport` but changing W leads to scaling (which hardware doesn't do). I also tried modifying `oPos` in the shader, but obviously it has the same issue (as W is still used by OpenGL).